src/Parsers/ExpressionListParsers.cpp

```diff
@@ -799,13 +801,52 @@ class ExpressionLayer : public Layer
         return Layer::getResultImpl(node);
     }
 
-    bool parse(IParser::Pos & pos, Expected & /*expected*/, Action & /*action*/) override
+    bool parse(IParser::Pos & pos, Expected & expected, Action & /*action*/) override
     {
         if (pos->type == TokenType::Comma)
+        {
             finished = true;
 
+            if (!allow_trailing_commas)
+                return true;
+
+            /// We support trailing commas at the end of the column declaration:
+            ///  - SELECT a, b, c, FROM table
+            ///  - SELECT 1,
+
+            /// For this purpose we eliminate the following cases:
+            ///  1. WITH 1 AS from SELECT 2, from
+            ///  2. SELECT to, from FROM table
+            ///  3. SELECT to, from AS alias FROM table
+            ///  4. SELECT to, from + to FROM table
+
+            auto test_pos = pos;
+            ++test_pos;
+
+            if (test_pos.isValid() && test_pos->type != TokenType::Semicolon)
+            {
+                if (!ParserKeyword("FROM").ignore(test_pos, expected))
+                    return true;
+
+                if (ParserKeyword("FROM").ignore(test_pos, expected))
```

Comment:
It's difficult to understand these two statements:
- if we didn't eat "FROM", return;
- if we ate "FROM", also return;
- then what are the next lines?

Reply:
1. If we didn't eat `FROM`, return;
2. If we eat two `FROM` in a row then the first `FROM` is probably an identifier for a column, return;
3. If we parse an explicit alias to `FROM`, return;
4. Here we are trying to (kind of) filter out the cases where `FROM` is followed by an operator. It is probably a good idea to search for all the operators in `ParserExpressionImpl::operators_table`, but then in cases when the table name is for example `between` the trailing comma will not be eaten.

Reply:
Good - let's add all of these as comments and merge.